vulnerabilities/importers/gitlab.py

```diff
@@ -0,0 +1,347 @@
+from univers.version_range import VersionRange
+from univers.version_range import from_gitlab_native
+from univers.versions import Version
+
+from vulnerabilities.helpers import AffectedPackage as LegacyAffectedPackage
+from vulnerabilities.helpers import nearest_patched_package
+from vulnerabilities.helpers import resolve_version_range
+from vulnerabilities.importer import AdvisoryData
+from vulnerabilities.importer import AffectedPackage
+from vulnerabilities.importer import Importer
+from vulnerabilities.importer import Reference
+from vulnerabilities.importer import UnMergeablePackageError
+from vulnerabilities.improver import Improver
+from vulnerabilities.improver import Inference
+from vulnerabilities.models import Advisory
+from vulnerabilities.package_managers import VERSION_API_CLASSES_BY_PACKAGE_TYPE
+from vulnerabilities.package_managers import GoproxyVersionAPI
+from vulnerabilities.package_managers import VersionAPI
+from vulnerabilities.package_managers import get_api_package_name
+
+logger = logging.getLogger(__name__)
+
+
+PURL_TYPE_BY_GITLAB_SCHEME = {
+    "gem": "gem",
+    "go": "golang",
+    "maven": "maven",
+    "npm": "npm",
+    "nuget": "nuget",
+    "pypi": "pypi",
+    "packagist": "composer",
+}
+
+
+GITLAB_SCHEME_BY_PURL_TYPE = {v: k for k, v in PURL_TYPE_BY_GITLAB_SCHEME.items()}
+
+
+def fork_and_get_dir(url):
+    """
+    Fetch a clone of the gitlab repository at url and return the directory destination
+    """
+    return fetch_via_vcs(url).dest_dir
+
+
+class ForkError(Exception):
+    pass
+
+
+class GitLabAPIImporter(Importer):
+    spdx_license_expression = "MIT"
+    license_url = "https://gitlab.com/gitlab-org/advisories-community/-/blob/main/LICENSE"
+    gitlab_url = "git+https://gitlab.com/gitlab-org/advisories-community/"
+
+    def advisory_data(self) -> Iterable[AdvisoryData]:
+        try:
+            fork_directory = fork_and_get_dir(self.gitlab_url)
+        except Exception as e:
+            logger.error(f"Can't clone url {self.gitlab_url}")
+            raise ForkError(self.gitlab_url) from e
+        ecosystems = ["nuget", "maven", "gem", "npm", "go", "packagist", "pypi"]
+        for ecosystem in ecosystems:
+            for file in get_files(os.path.join(fork_directory, ecosystem)):
+                yield parse_yaml_file(file)
+
+
+def get_files(dir):
+    for root, _, files in os.walk(dir):
+        for file in files:
+            yield os.path.join(root, file)
+
+
+def not_empty(value):
+    return value is not None and value != ""
+
+
+def get_purl(package_slug):
+    """
+    Return a PackageURL object from a package slug
+    """
+    parts = package_slug.split("/")
+    parts = list(filter(not_empty, parts))
+    gitlab_scheme = parts[0]
+    purl_type = PURL_TYPE_BY_GITLAB_SCHEME[gitlab_scheme]
+    if gitlab_scheme == "go":
+        name = "/".join(parts[1:])
+        return PackageURL(type=purl_type, namespace=None, name=name)
```

Comment:
I am following the same convention that is used in github importer to store package name in go package type (We don't store namespace in github and only depends on name to fetch data from API)

Reply:
As an example I think that this slug `go/gitlab.com/gitlab-org/gitlab-shell` would become this purl: `pkg:golang/gitlab.com/gitlab-org/gitlab-shell` and if you parse it you get this:
```Python
>>> PackageURL.from_string("pkg:golang/gitlab.com/gitlab-org/gitlab-shell")
PackageURL(type='golang', namespace='gitlab.com/gitlab-org', name='gitlab-shell', version=None, qualifiers={}, subpath=None)
```
But this is incorrect:
```Python
>>> PackageURL(type='golang', namespace=None, name='gitlab.com/gitlab-org/gitlab-shell')
PackageURL(type='golang', namespace=None, name='gitlab.com/gitlab-org/gitlab-shell', version=None, qualifiers={}, subpath=None)
```
.... even if this may look correct on the surface:
```Python
>>> PackageURL(type='golang', namespace=None, name='gitlab.com/gitlab-org/gitlab-shell').to_string()
'pkg:golang/gitlab.com/gitlab-org/gitlab-shell'
```

IMHO this would be much safer and clean
```Python
>>> slug = "go/gitlab.com/gitlab-org/gitlab-shell"
>>> scheme, _, ns_name = slug.partition("/")
>>> assert scheme == "go"
>>> purl = f"pkg:golang/{ns_name}"
>>> purl
'pkg:golang/gitlab.com/gitlab-org/gitlab-shell'
>>> PackageURL.from_string(purl)
PackageURL(type='golang', namespace='gitlab.com/gitlab-org', name='gitlab-shell', version=None, qualifiers={}, subpath=None)

```

Reply:
refer #742 for followup on this